src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/HttpGooglePubSubDAO.scala

```diff
@@ -92,30 +64,47 @@ class HttpGooglePubSubDAO(
     }})
   }
 
-  def acknowledgeMessage(subscriptionName: String, ackId: String) = {
+  def publishMessages(topicName: String, messages: Seq[String]) = {
+    retry(when500)(() => Future { blocking {
+      val pubsubMessages = messages.map(text => new PubsubMessage().encodeData(text.getBytes("UTF-8")))
+      val pubsubRequest = new PublishRequest().setMessages(pubsubMessages)
+      executeGoogleRequest(getPubSubDirectory.projects().topics().publish(topicToFullPath(topicName), pubsubRequest))
+    }})
+  }
+
+  def acknowledgeMessages(subscriptionName: String, ackIds: Seq[String]) = {
     retry(when500)(() => Future { blocking {
-      val ackRequest = new AcknowledgeRequest().setAckIds(List(ackId))
+      val ackRequest = new AcknowledgeRequest().setAckIds(ackIds)
       executeGoogleRequest(getPubSubDirectory.projects().subscriptions().acknowledge(subscriptionToFullPath(subscriptionName), ackRequest))
     }})
   }
 
-  def pullMessage(subscriptionName: String): Future[Option[ReceivedMessage]] = {
+  def pullMessages(subscriptionName: String, maxMessages: Int): Future[Seq[ReceivedMessage]] = {
     retry(when500)(() => Future { blocking {
-      val pullRequest = new PullRequest().setReturnImmediately(true).setMaxMessages(1) //won't keep the connection open if there's no msgs available
+      val pullRequest = new PullRequest().setReturnImmediately(true).setMaxMessages(maxMessages) //won't keep the connection open if there's no msgs available
       val messages = executeGoogleRequest(getPubSubDirectory.projects().subscriptions().pull(subscriptionToFullPath(subscriptionName), pullRequest)).getReceivedMessages
       if(messages == null)
-        None
-      else messages.headOption
+        Seq.empty
+      else messages.toList.toSeq
     }})
   }
 
-  def withMessage(subscriptionName: String)(op: (String) => Future[AckStatus]): Future[HandledStatus] = {
-    pullMessage(subscriptionName) flatMap {
-      case Some(msg) => op(decodeMessage(msg)) flatMap {
-        case MessageAcknowledged => acknowledgeMessage(subscriptionName, msg.getAckId).map(_ => MessageHandled) //HandledStatus: msg handled
-        case MessageNotAcknowledged => Future.successful(MessageNotHandled) //HandledStatus: msg unhandled
+  def withMessage(subscriptionName: String)(op: (String) => Future[AckStatus.Value]): Future[HandledStatus.Value] = {
```

Comment:
why isn't this
withMessages(subscriptionName, 1) {
  case Seq(msg) => op(msg)
  case _ => exception
}